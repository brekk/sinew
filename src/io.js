import fs from "graceful-fs"
import { node } from "fluture"
import { curry, curryN, __ as $, pipe } from "ramda"

// async - future-producing

/**
 * fs.readFile but curried
 * @method readRaw
 * @public
 * @param {string} path - path-to-a-file
 * @param {string} format - file format
 * @param {function} callback - standard nodeback style callback
 * @return null
 * @example
 * import { readRaw } from "sinew"
 * import { pipe, map, __ as skip } from "ramda"
 * import { node } from "fluture"
 *
 * export readFileAndDoStuff = pipe(
 *   pipe(
 *     readRaw(skip, 'utf8'),
 *     node
 *   ),
 *   map(x => "preamble:\n" + x)
 * )
 */
export const readRaw = curryN(3, fs.readFile)
/**
 * fs.readFile but curried and with utf8 format chosen
 * @method readUTF8
 * @public
 * @param {string} path - path-to-a-file
 * @param {function} callback - a node-style nodeback function
 * @return {string} a future value of a string
 * @example
 * import { readUTF8 } from "sinew"
 * import { pipe, map, __ as skip } from "ramda"
 * import { node } from "fluture"
 *
 * export readFileAndDoStuff = pipe(
 *   pipe(
 *     readUTF8(skip),
 *     node
 *   ),
 *   map(x => "preamble:\n" + x)
 * )
 */
export const readUTF8 = readRaw($, `utf8`)

/**
 * fs.readFile but utf8 and returning a Future
 * @method readFile
 * @public
 * @param {string} path - path-to-a-file
 * @return {Future<string>} a future value of a string
 * @example
 * import { readFile } from "sinew"
 * import { pipe, map } from "ramda"
 *
 * export readFileAndDoStuff = pipe(
 *   readFile,
 *   map(x => "preamble:\n" + x)
 * )
 */
export const readFile = pipe(
  readUTF8,
  node
)

/**
 * fs.readFile but curried; arity 4
 * @method writeRaw
 * @public
 * @param {string} path - path-to-a-file
 * @param {*} data - data to write to a path
 * @param {string|object} format - format or opts
 * @param {function} callback - a nodeback-style callback function
 * @example
 * import {writeRaw} from "sinew"
 * writeRaw("my-file.md", "cool", "utf8", (e) => {
 *   if(e) console.warn(e)
 *   // done
 * })
 */
export const writeRaw = curryN(4, fs.writeFile)

/**
 * fs.readFile but curried; format utf8
 * @method writeRaw
 * @public
 * @param {string} path - path-to-a-file
 * @param {*} data - data to write to a path
 * @param {function} callback - a nodeback-style callback function
 * @example
 * import { writeUTF8 } from "sinew"
 *
 * writeUTF8('my-file.md', 'cool', (e) => {
 *   if(e) console.warn(e)
 *   // done
 * })
 */
export const writeUTF8 = writeRaw($, $, `utf8`)

/**
 * Write a utf8 file and wrap the action in a future
 * @method writeFile
 * @public
 * @param {string} to - path-to-a-file
 * @param {string} data - data to write
 * @return {Future<string>} a future-wrapped value(?)
 * @example
 * import { writeFile, readFile } from "sinew"
 * import { curry, pipe, map, chain } from "ramda"
 *
 * const prepend = curry((pre, file, data) => pipe(
 *   readFile(file),
 *   map(raw => pre = raw),
 *   chain(writeFile)
 * ))
 */
export const writeFile = curry((to, data) =>
  pipe(
    writeUTF8(to),
    node
  )(data)
)
