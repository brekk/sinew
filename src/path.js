import path from "path"
import { curryN } from "ramda"

export const relativePathWithCWD = curryN(2, path.relative)
