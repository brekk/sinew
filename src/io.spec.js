/* global test, expect */
import path from "path"
import { readFile } from "./io"
const here = (...x) => path.join(__dirname, ...x)

const jsonPkg = here("..", "package.json")

test("readFile", done => {
  readFile(jsonPkg).fork(done, raw => {
    expect(JSON.parse(raw)).toMatchSnapshot()
    done()
  })
})
