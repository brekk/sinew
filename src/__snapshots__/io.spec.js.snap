// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`readFile 1`] = `
Object {
  "author": "brekk",
  "dependencies": Object {
    "cliui": "5.0.0",
    "debug": "4.1.1",
    "execa": "1.0.0",
    "fluture": "11.0.1",
    "get-stdin": "7.0.0",
    "kleur": "3.0.3",
    "ramda": "0.26.1",
    "strip-ansi": "5.2.0",
  },
  "description": "",
  "devDependencies": Object {
    "@babel/cli": "7.4.4",
    "@babel/core": "7.4.5",
    "@babel/plugin-transform-destructuring": "7.4.4",
    "@babel/preset-env": "7.4.5",
    "babel-core": "6.26.3",
    "babel-eslint": "10.0.1",
    "babel-jest": "24.8.0",
    "depcheck": "0.6.11",
    "documentation": "13.1.1",
    "eslint": "5.16.0",
    "eslint-config-prettier": "4.3.0",
    "eslint-config-standard": "12.0.0",
    "eslint-plugin-babel": "5.3.0",
    "eslint-plugin-fp": "2.3.0",
    "eslint-plugin-import": "2.17.3",
    "eslint-plugin-node": "9.1.0",
    "eslint-plugin-prettier": "3.1.0",
    "eslint-plugin-promise": "4.1.1",
    "eslint-plugin-ramda": "2.5.1",
    "eslint-plugin-standard": "4.0.0",
    "nps": "5.9.5",
    "nps-utils": "1.7.0",
    "prettier-eslint": "8.8.2",
    "rollup": "1.13.1",
    "rollup-plugin-babel": "4.3.2",
    "rollup-plugin-cleanup": "3.1.1",
    "rollup-plugin-commonjs": "10.0.0",
    "rollup-plugin-json": "4.0.0",
    "rollup-plugin-node-resolve": "5.0.1",
    "rollup-plugin-progress": "1.1.1",
    "tempy": "0.3.0",
    "xtrace": "0.3.0",
  },
  "keywords": Array [],
  "license": "MIT",
  "main": "sinew.js",
  "module": "sinew.es.js",
  "name": "sinew",
  "private": false,
  "repository": "https://github.com/brekk/sinew",
  "version": "1.0.2",
}
`;
